### ISSUE TITLE
[QUESTION] Is there a way to use Hashicorp Vault that stores the CRYPTO_KEY_VALUE when using encryption

### ISSUE THREAD

**Issue by younesehb** [longhorn/longhorn#3026]
**Question**
Is there a way to use Hashicorp Vault that stores the CRYPTO_KEY_VALUE when using encryption ? 

**Environment:**
 - Longhorn version: v1.2.0
 - Kubernetes version: v1.21.3
 - Node config
   - OS type and version: Ubuntu 20.04.2 LTS, kernel version 5.4.0-72-generic
   - CPU per node: 4 CPU
   - Memory per node: 8 GB
   - Disk type: SSD
   - Network bandwidth and latency between the nodes:
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Exoscale cloud

**Additional context**
When creating the Kubernetes secret that will be used by Longhorn to encrypt persistent volumes, the secret is in clear text.
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: longhorn-crypto
  namespace: longhorn-system
stringData:
# This needs to be encrypted using the Vault later
  CRYPTO_KEY_VALUE: "C_Fm8#$3@x"
  CRYPTO_KEY_PROVIDER: "secret" # this is optional we currently only support direct keys via secrets 
```
I would like to use Hashicorp Vault to avoid having a cleartext value here.

**Comment by jenting** [comment]
Related to #2804

**Comment by snagiel** [comment]
Vault already provides mechanisms (and tutorials) on utilizing its secret engine to:
1) [inject values into pods with the agent container sidecar](https://learn.hashicorp.com/tutorials/vault/kubernetes-sidecar?in=vault/kubernetes)
2) [mount (Vault-managed) secrets as proxied volumes](https://learn.hashicorp.com/tutorials/vault/kubernetes-secret-store-driver?in=vault/kubernetes)

I would think that either one of these approaches could work with a Longhorn-configured encrypted volume (i.e. rather than requiring a k8s secret object with the `CRYPTO_KEY_VALUE` key).

**Comment by jenting** [comment]
Thank you for the information.
We'll investigate to see how longhorn to achieve it with Vault.

**Comment by snagiel** [comment]
In addition to the two options enumerated above, Vault also provides the [Kubernetes Auth Method](https://www.vaultproject.io/docs/auth/kubernetes), which [Rook/Ceph offer as a KMS solution for storing dmcrypt keys](https://rook.io/docs/rook/v1.8/ceph-kms.html). I think this is likely the cleanest approach.

Is there any timetable for tackling this ticket?

**Comment by jenting** [comment]
> In addition to the two options enumerated above, Vault also provides the [Kubernetes Auth Method](https://www.vaultproject.io/docs/auth/kubernetes), which [Rook/Ceph utilizes as a KMS solution for storing dmcrypt keys](https://rook.io/docs/rook/v1.8/ceph-kms.html). I think this is likely the cleanest approach.
> 
> Is there any timetable for tackling this ticket?

cc @innobead

**Comment by innobead** [comment]
No recent schedule, but this will be planned in the following feature releases.

**Comment by joshimoo** [comment]
When I designed this, I had future extensibility in mind which you can see by the fact that there is a `CRYPTO_KEY_PROVIDER` field in the secret which allows us to implement other providers while utilizing the same kubernetes -> csi sidecars -> csi plugin secret mechanism, which allows us then to do the resolving of the final `CRYPTO_KEY` as part of the csi-plugin.

TLDR: it's technically feasible for us to add support for additional key management solutions (hashicorp, aws, etc) in future versions.

**Comment by jenting** [comment]
The design could consider the [SecretProviderClass](https://secrets-store-csi-driver.sigs.k8s.io/getting-started/usage.html).

**Comment by boopzz** [comment]
Could we add in scope to manage any of the potential secrets for longhorn? ie S3 API keys for backups

I've just setup s3 backups (to Linode) and feel like it would be better to have those secrets in Vault and injected into the pod. Feels like very similar functionality?

I can add another issue if it needs separate tracking?

**Comment by joshimoo** [comment]
@boopzz I think it's a valid ask to also have these mechanisms for any other type of Keys (Backups, ssl certs, etc). But since they require completely different implementations internally, It would be appreciated if you open a separate enhancement request with your use case. So it can be tracked and prioritized independently.
